Sources/SymbolKit/SymbolGraph/Symbol/AlternateDeclarations.swift

```diff
@@ -36,23 +37,147 @@ extension SymbolGraph.Symbol {
+            self.alternateSymbols = alternateDeclarations.declarations.map { declaration in
+                .init(declarationFragments: declaration)
+            }
+        }
+
+        /// The list of alternate symbol information.
+        public var alternateSymbols: [AlternateSymbol]
+    }
+
     /// Convenience accessor to fetch alternate declarations for this symbol.
     public var alternateDeclarations: [DeclarationFragments]? {
-        (mixins[AlternateDeclarations.mixinKey] as? AlternateDeclarations)?.declarations
+        alternateSymbols?.alternateSymbols.compactMap(\.declarationFragments)
     }
 
-    internal mutating func addAlternateDeclaration(_ declaration: DeclarationFragments) {
-        if var alternateDeclarations = mixins[AlternateDeclarations.mixinKey] as? AlternateDeclarations {
-            alternateDeclarations.declarations.append(declaration)
-            mixins[AlternateDeclarations.mixinKey] = alternateDeclarations
-        } else {
-            mixins[AlternateDeclarations.mixinKey] = AlternateDeclarations(declarations: [declaration])
-        }
+    /// Convenience accessor to fetch alternate symbol information.
+    ///
+    /// Returns `nil` if there were no alternate symbols found when decoding the symbol graph.
+    public var alternateSymbols: AlternateSymbols? {
+        mixins[AlternateSymbols.mixinKey] as? AlternateSymbols
+    }
+
+    /// Convenience accessor to fetch a specific mixin from this symbol's alternate symbols.
+    ///
+    /// Because the mixin key is inferred from the type parameter,
+    /// the easiest way to call this method is by assigning it to a variable with an appropriate type:
+    ///
+    /// ```swift
+    /// let alternateFunctionSignatures: [SymbolGraph.Symbol.FunctionSignature]? = symbol.alternateSymbolMixins()
+    /// ```
+    ///
+    /// If there are multiple alternate symbols and only some of them have the requested mixin,
+    /// any missing data is removed from the resulting array via a `compactMap`.
+    public func alternateSymbolMixins<M: Mixin>(mixinKey: String = M.mixinKey) -> [M]? {
+        alternateSymbols?.alternateSymbols.compactMap({ $0.mixins[mixinKey] as? M })
     }
 
     internal mutating func addAlternateDeclaration(from symbol: SymbolGraph.Symbol) {
-        if let declaration = symbol.mixins[DeclarationFragments.mixinKey] as? DeclarationFragments {
-            addAlternateDeclaration(declaration)
+        let alternate = AlternateSymbols.AlternateSymbol(symbol: symbol)
+        guard !alternate.isEmpty else { return }
+
+        if var alternateSymbols = mixins[AlternateSymbols.mixinKey] as? AlternateSymbols {
+            alternateSymbols.alternateSymbols.append(alternate)
+            mixins[AlternateSymbols.mixinKey] = alternateSymbols
+        } else {
+            mixins[AlternateSymbols.mixinKey] = AlternateSymbols(alternateSymbols: [alternate])
         }
     }
 }
+
+extension SymbolGraph.Symbol.AlternateSymbols {
```

Comment:
Question: why is this extending a new type to add a deprecated type property instead of deprecating and using `AlternateDeclarations.symbolCodingInfo`?

Reply:
I wanted to decode `alternateDeclarations` into the new `alternateSymbols` type, and i wasn't sure about how to do that in the regular `Symbol` decode method. I could take another look and see if i can special-case it.